### PR TITLE
:sparkles: PIC-496: Add current order header details

### PR DESCRIPTION
**Pull request by katie-cawthorne**
- Added community-api call `/secure/offenders/crn/{crn}/convictions/{convictionId/sentences/{sentenceId}/custodialStatus`
- Updated conviction endpoint to bring in sentenceId needed for call to new community-api call
- Renamed some functions to refer to 'sentence' instead of conviction
- Added currentOrderHeaderDetail data to modified conviction endpoint

### CODE REVIEW

File: src/main/java/uk/gov/justice/probation/courtcaseservice/controller/model/CurrentOrderHeaderResponse.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude;
+import io.swagger.annotations.ApiModel;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import uk.gov.justice.probation.courtcaseservice.service.model.KeyValue;
+
+import java.time.LocalDate;
+
+@ApiModel(description = "Current order header detail")
```
Comment (chris-faulkner-mojdt) [inline src/main/java/uk/gov/justice/probation/courtcaseservice/controller/model/CurrentOrderHeaderResponse.java:13]: We've gone down the route of enforcing immutability on our model objects. I've done it on the PR parallel to this but it basically involves making fields final. I think you can also use the Value lombok annotation

File: src/main/java/uk/gov/justice/probation/courtcaseservice/controller/model/SentenceResponse.java [hunk 2]
```diff
@@ -11,17 +11,25 @@
...
 
-@ApiModel(description = "Conviction Response")
+@ApiModel(description = "Sentence Response")
 @Data
 @Builder
 @AllArgsConstructor
 @NoArgsConstructor
 @JsonInclude(Include.NON_NULL)
-public class ConvictionResponse {
+public class SentenceResponse {
```
Comment (chris-faulkner-mojdt) [inline src/main/java/uk/gov/justice/probation/courtcaseservice/controller/model/SentenceResponse.java:20]: See previous comment about immutability, although I think we are going to run into merge issues :-)

File: src/main/java/uk/gov/justice/probation/courtcaseservice/restclient/communityapi/mapper/OffenderMapper.java [hunk 3]
```diff
@@ -7,21 +7,9 @@
...
-import uk.gov.justice.probation.courtcaseservice.service.model.Conviction;
-import uk.gov.justice.probation.courtcaseservice.service.model.Offence;
-import uk.gov.justice.probation.courtcaseservice.service.model.OffenderManager;
-import uk.gov.justice.probation.courtcaseservice.service.model.ProbationRecord;
-import uk.gov.justice.probation.courtcaseservice.service.model.Requirement;
-import uk.gov.justice.probation.courtcaseservice.service.model.Sentence;
-import uk.gov.justice.probation.courtcaseservice.service.model.UnpaidWork;
+import uk.gov.justice.probation.courtcaseservice.controller.model.CurrentOrderHeaderResponse;
+import uk.gov.justice.probation.courtcaseservice.restclient.communityapi.model.*;
+import uk.gov.justice.probation.courtcaseservice.service.model.*;
```
Comment (chris-faulkner-mojdt) [inline src/main/java/uk/gov/justice/probation/courtcaseservice/restclient/communityapi/mapper/OffenderMapper.java:12]: There was a review comment on Community API from DPS team that they had stopped wildcard imports. Perhaps we could follow suit.

File: src/main/java/uk/gov/justice/probation/courtcaseservice/restclient/communityapi/model/CommunityApiCurrentOrderHeaderDetailResponse.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+import java.time.LocalDate;
+
+@ApiModel(description = "Current order header detail response from Community API")
+@Getter
+@AllArgsConstructor
+@Builder
+@NoArgsConstructor
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class CommunityApiCurrentOrderHeaderDetailResponse {
```
Comment (chris-faulkner-mojdt) [inline src/main/java/uk/gov/justice/probation/courtcaseservice/restclient/communityapi/model/CommunityApiCurrentOrderHeaderDetailResponse.java:18]: See comment on immutability

File: src/test/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderControllerTest.java [hunk 5]
```diff
@@ -64,27 +70,37 @@ void beforeEach() {
...
-    void featureToggleFalse() {
-        featureFlags.setFlagValue("fetch-attendance-data", false);
-        final ConvictionResponse convictionResponse = ConvictionResponse.builder()
+    void featureToggleFalseSentenceData() {
+        featureFlags.setFlagValue("fetch-sentence-data", false);
+        final SentenceResponse sentenceResponse = SentenceResponse.builder()
                 .attendances(Collections.emptyList())
                 .unpaidWork(UnpaidWork.builder().build())
+                .currentOrderHeaderDetail(CurrentOrderHeaderResponse.builder().custodialType(KeyValue.builder().code(null).description(null).build())
+                        .sentenceDescription(null)
```
Comment (chris-faulkner-mojdt) [inline src/test/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderControllerTest.java:90]: might be easier if we just didn't call all these will null. They'll be null without that.

File: src/main/java/uk/gov/justice/probation/courtcaseservice/restclient/communityapi/mapper/OffenderMapper.java (hunk 3 above)
Comment (katie-cawthorne) [reply src/main/java/uk/gov/justice/probation/courtcaseservice/restclient/communityapi/mapper/OffenderMapper.java:12]: Yes I agree, I would prefer to not use wildcard imports by choice but intellij sneaked it in for me. I'll change 🙂

File: src/test/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderControllerTest.java (hunk 5 above)
Comment (katie-cawthorne) [reply src/test/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderControllerTest.java:90]: Good spot

File: src/main/java/uk/gov/justice/probation/courtcaseservice/application/FeatureFlags.java [hunk 6]
```diff
@@ -19,8 +19,8 @@ public FeatureFlags() {
         return flags;
     }
 
-    public boolean attendanceData() {
-        return flags.getOrDefault("fetch-attendance-data", true);
```
Comment (chris-faulkner-mojdt) [inline src/main/java/uk/gov/justice/probation/courtcaseservice/application/FeatureFlags.java:23]: Has fetch-attendance-data gone completely ? Was this flag supposed to just get rid of attendance data ?

Comment (chris-faulkner-mojdt) [reply src/main/java/uk/gov/justice/probation/courtcaseservice/application/FeatureFlags.java:23]: If so, could you get rid of all the occurrences of that flag ? It's still in some YAML and test code

File: src/test/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderController_ConvictionIntTest.java [hunk 7]
```diff
@@ -38,7 +40,7 @@
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT, properties = "org.apache.catalina.connector.RECYCLE_FACADES=true")
 public class OffenderController_ConvictionIntTest {
```
Comment (chris-faulkner-mojdt) [inline src/test/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderController_ConvictionIntTest.java:41]: In setUp, it's setting featureFlags.setFlagValue("fetch-attendance-data",true) - is that used ?

File: src/main/java/uk/gov/justice/probation/courtcaseservice/service/ConvictionService.java [hunk 8]
```diff
@@ -6,42 +6,68 @@
...
-import uk.gov.justice.probation.courtcaseservice.service.model.Conviction;
-import uk.gov.justice.probation.courtcaseservice.service.model.Sentence;
+import uk.gov.justice.probation.courtcaseservice.service.model.*;
 
 @Service
 public class ConvictionService {
 
     private final ConvictionRestClient restClient;
 
+    private CurrentOrderHeaderResponse currentOrderHeaderResponse = new CurrentOrderHeaderResponse(null, null, null, null, null, null, null, null, null, null);
```
Comment (chris-faulkner-mojdt) [inline src/main/java/uk/gov/justice/probation/courtcaseservice/service/ConvictionService.java:23]: This is unused and can be removed

File: src/main/java/uk/gov/justice/probation/courtcaseservice/service/model/CurrentOrderHeaderDetail.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+import lombok.Getter;
+
+import java.time.LocalDate;
+
+@ApiModel("CurrentOrderHeaderDetail")
+@Getter
+@Builder
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class CurrentOrderHeaderDetail {
+    @JsonIgnore
```
Comment (chris-faulkner-mojdt) [inline src/main/java/uk/gov/justice/probation/courtcaseservice/service/model/CurrentOrderHeaderDetail.java:16]: Wht do we not return this field  ? Would prepare-a-case need it in future requests ?

File: src/test/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderController_ConvictionIntTest.java [hunk 10]
```diff
@@ -101,32 +105,47 @@ public void whenCallMadeToGetConvictionKnownCrnAndConvictionId() {
                                                                                             .unacceptableAbsences(1)
                                                                                             .status("Being worked")
                                                                                             .build());
+
+        assertThat(response.getCurrentOrderHeaderDetail()).isEqualToComparingFieldByField(CurrentOrderHeaderResponse.builder()
+//                .sentenceId(2500298861L)
```
Comment (chris-faulkner-mojdt) [inline src/test/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderController_ConvictionIntTest.java:110]: Just delete this line

File: src/main/java/uk/gov/justice/probation/courtcaseservice/service/model/CurrentOrderHeaderDetail.java (hunk 9 above)
Comment (katie-cawthorne) [reply src/main/java/uk/gov/justice/probation/courtcaseservice/service/model/CurrentOrderHeaderDetail.java:16]: I wasn't sure whether to have the sentenceId field inside the currentOrderHeaderDetail or returned as a field as part of the SentenceResponse, so I ignored it from the CurrentOrderHeaderDetail and included the sentenceId as a field as part of the SentenceResponse instead, if you see what I mean.

File: src/main/java/uk/gov/justice/probation/courtcaseservice/service/ConvictionService.java (hunk 8 above)
Comment (katie-cawthorne) [reply src/main/java/uk/gov/justice/probation/courtcaseservice/service/ConvictionService.java:23]: Good spot I was supposed to delete that

**Pull request merged**